docs/docs/refs-and-the-dom.md

```diff
@@ -0,0 +1,62 @@
+---
+id: refs-and-the-dom
+title: Refs and the DOM
+permalink: docs/refs-and-the-dom.html
+---
+
+In the typical React dataflow, `props` are the only way that parent components interact with their children. To modify a child, you re-render it with new `props`. However, there are a few cases where you need to imperatively modify a child outside of the typical dataflow. The child to be modified could be an instance of a React component, or it could be a DOM element. For both of these cases, React provides an escape hatch.
+
+## The ref Callback Attribute
+
+React supports a special attribute that you can attach to any component. The `ref` attribute takes a callback function, and the callback will be executed immediately after the component is mounted.
+
+When the `ref` attribute is used on a HTML element, the `ref` callback receives the underlying DOM element as its argument. For example, this code uses the `ref` callback to store a reference to a DOM node:
+
+```javascript
+class CustomTextInput extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleClick = this.handleClick.bind(this);
+  }
+
+  handleClick() {
+    // Explicitly focus the text input using the raw DOM API
+    if (this.textInput !== null) {
```

Comment:
This encourages defensive programming, doesn't it? We know that it is always guaranteed to be mounted because `input` is rendered together with `button`. We'd need this check if `input` was conditionally rendered but it seems unnecessary here.